lib/opam.ml

```diff
@@ -238,3 +238,29 @@ module Pos = struct
     let pos = from_value value in
     errorf ~pos fmt
 end
+
+module Value = struct
+  module String = struct
+    let from_value value =
+      match (value : OpamParserTypes.FullPos.value) with
+      | { pelem = String s; _ } -> Ok s
+      | { pos; _ } -> Pos.errorf ~pos "Expected a string"
```

Comment:
Given this function is now reusable, maybe the error message could specify what it got instead of the string? That could make it easier for users who run into this to catch which part of the OPAM file caused the error.

Something like

```ocaml
| { pelem; pos; } -> Pos.errorf ~pos "Expected a string, got %a" pp_pelem pelem
```

so that `pp_pelem` could also be used in `Opam.Value.from_value`?

Reply:
Note that the error is reported along with the location so I'm not sure there is a lot of value in describing what is there.

We could further improve this error reporting to pretty print the surrounding parts of the file, a la OCaml compiler but this feels a bit overkill for now I think.

To calrify what I have in mid would be something like:
```
In opam file foo.opam, line 7, char 5 - 23: Expected a string
...
6-    "A proper string
7-    [ "A String wrapped in a list by mistake"]
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
```

If you're okay with this, I suggest we keep this PR a refactoring and in the future if we receive complaints with error message in opam files extensions parsing we can eventually put some effort into improving them. If you do feel like the current error messages are specifically bad, we can eventually improve it in a subsequent PR as suggested above.

Reply:
I agree that doing a proper error reporting with syntax is way beyond the scope of this PR: this is complex stuff and I would even argue, not particularly necessary here. Lots of work for little reward.

I still think that that printing what the program sees here would be quite easy to do. I find it useful e.g. when the OCaml compiler tells me what type it expected vs what type it got. But admittedly this can also easily done later so I'd say it is up to you, I don't consider this a must-have feature for this PR :-)

Reply:
Added just that!